lib/hsa/hsaAsyncCopyGain.cpp

```diff
@@ -0,0 +1,105 @@
+#include "hsaAsyncCopyGain.hpp"
+
+#include "utils/util.h"
+
+#include "fmt.hpp"
+
+#include <random>
+#include <utils/visUtil.hpp>
+
+using kotekan::bufferContainer;
+using kotekan::Config;
+
+REGISTER_HSA_COMMAND(hsaAsyncCopyGain);
+
+hsaAsyncCopyGain::hsaAsyncCopyGain(Config& config, const string& unique_name,
+                                   bufferContainer& host_buffers, hsaDeviceInterface& device) :
+    hsaCommand(config, unique_name, host_buffers, device, "hsaAsyncCopyGain", "") {
+    command_type = gpuCommandType::COPY_IN;
+
+    gain_len = 2 * 2048 * sizeof(float);
+    gain_buf = host_buffers.get_buffer("gain_frb_buf");
+    register_consumer(gain_buf, unique_name.c_str());
+    gain_buf_id = 0;
+    gain_buf_finalize_id = 0;
+    gain_buf_precondition_id = 0;
+    frame_to_fill = 0;
+    frame_to_fill_finalize = 0;
+    filling_frame = false;
+    first_pass = true;
+}
+
+
+hsaAsyncCopyGain::~hsaAsyncCopyGain() {}
+
+int hsaAsyncCopyGain::wait_on_precondition(int gpu_frame_id) {
+    (void)gpu_frame_id;
+
+    // Check for new gains
+    DEBUG("Waiting for gain_buf_id={:d} to be full; gpu_frame_id={:d}", gain_buf_precondition_id,
+          gpu_frame_id);
+    if (first_pass) {
+        uint8_t* frame =
+            wait_for_full_frame(gain_buf, unique_name.c_str(), gain_buf_precondition_id);
+        gain_buf_precondition_id = (gain_buf_precondition_id + 1) % gain_buf->num_frames;
+        first_pass = false;
+        frame_to_fill = gain_buf->num_frames;
+        frame_to_fill_finalize = frame_to_fill;
+        filling_frame = true;
+        if (frame == NULL)
+            return -1;
+    } else {
+        // Check for new gains only if filled all gpu frames (not currently filling frame)
+        if (!filling_frame) {
+            auto timeout = double_to_ts(0);
+            int status = wait_for_full_frame_timeout(gain_buf, unique_name.c_str(),
+                                                     gain_buf_precondition_id, timeout);
+            DEBUG("status of gain_buf_precondition_id[{:d}]={:d} ==(0=ready 1=not)================",
+                  gain_buf_precondition_id, status);
+            if (status == 0) {
+                filling_frame = true;
+                frame_to_fill = gain_buf->num_frames;
```

Comment:
You just need to fill all the GPU frames. 
```suggestion
                frame_to_fill = device.get_gpu_buffer_depth();
```